daemon/src/lib.rs

```diff
@@ -0,0 +1,655 @@
+/// parent will hang until one or the other occurs.)
+pub struct Daemonizing {
+    options: Daemonize,
+    notify_pipe: Option<File>
+}
+
+impl Daemonizing {
+    /// Completes the daemonization process:
+    ///
+    ///   - If we detached, changes the current directory to the root, unless that behavior is
+    ///     disabled.
+    ///
+    ///   - If we detached, closes the `stdin`, `stdout`, and `stderr` fds.  For safety, instead
+    ///     of leaving fds 0, 1, and 2 unpopulated, we replace them by `/dev/null`.
+    ///
+    ///   - Notifies the parent process that daemonization is complete.  This allows the parent
+    ///     process to exit successfully, indicating to the process that in turn
+    pub fn finish(mut self) {
+        if self.options.detach {
+            if !self.options.no_chdir {
+                drop(set_current_dir("/"));
+            }
+            close_standard_fds();
+        }
+        if let Some(ref mut pipe) = self.notify_pipe {
+            fork_notify_startup(pipe);
+        }
+    }
+
+    unsafe fn new(options: Daemonize) -> (Self, Cleanup) {
+        assert_single_threaded();
+
+        let mut notify_pipe = None;
+        if options.detach {
+            notify_pipe = match fork_and_wait_for_startup() {
+                ForkAndWaitResult::ForkFailed { status, .. } => {
+                    event!(Level::ERROR, "could not detach from foreground session ({status})");
+                    exit(1);
+                },
+                ForkAndWaitResult::InParent { .. } => exit(0),
+                ForkAndWaitResult::InChild { notify_pipe } => Some(notify_pipe),
+            };
+
+            // Running in daemon or monitor process.
+            libc::setsid();
+        }
+
+        if options.monitor {
+            notify_pipe = Some(match fork_and_wait_for_startup() {
+                ForkAndWaitResult::ForkFailed { status, .. } => {
+                    event!(Level::ERROR, "could not initiate process monitoring ({status})");
+                    exit(1);
+                },
+                ForkAndWaitResult::InParent { child_pid } => {
+                    // Running in monitor process.
+                    if let Some(ref mut notify_pipe) = notify_pipe {
+                        fork_notify_startup(notify_pipe);
+                    }
+                    if options.detach {
+                        close_standard_fds();
+                    }
+                    Self::monitor_daemon(child_pid)
+                },
+                ForkAndWaitResult::InChild { notify_pipe } => notify_pipe
+            })
+            // Running in daemon process.
+        }
+
+        // Running in daemon process.
+        proctitle::restore();
+        let mut cleanup = match cleanup::Cleanup::new() {
+            Ok(cleanup) => cleanup,
+            Err(error) => {
+                event!(Level::ERROR, "could not arrange for cleanup on process exit ({error})");
```

Comment:
so where do these logging messages go?

Reply:
Wherever the client program configured them to go. I did add a little comment about this to the end of the top-level module comment:

```
/// About all a program should do beforehand is to parse command-line options (needed to determine
/// whether to detach) and initialize logging (because daemonization will emit log entries).
```